Mathlib/Topology/Compactness/KSpace.lean

```diff
@@ -0,0 +1,315 @@
+    [SequentialSpace X] : KSpace X where
+  restrictGenTopology := RestrictGenTopology.isCompact_of_seq
+
+/-- In a k-space `X`, a set `s` is open when `f ⁻¹' s` is open for every continuous map from a
+compact space. -/
+lemma preimage_isOpen.{u} {X : Type u} [TopologicalSpace X] [KSpace X] (s : Set X)
+    (hs : (∀ (K : Type u) [TopologicalSpace K], [CompactSpace K] →
+      ∀ (f : K → X), Continuous f → IsOpen (f ⁻¹' s))) :
+    IsOpen s := by
+  rw [isOpen_iff]
+  intro K hK
+  let _ : CompactSpace (Elem K) := isCompact_iff_compactSpace.mp hK
+  exact hs (Elem K) Subtype.val continuous_subtype_val
+
+/-- A topological space `X` is compactly generated if a set `s` is open when `f ⁻¹' s?` is open for
+every continuous map `f : K → X`, where `K` is compact. -/
+lemma kspace_of_preimage_isOpen.{u} {X : Type u} [TopologicalSpace X]
+    (h : ∀ (s : Set X), (∀ (K : Type u) [TopologicalSpace K], [CompactSpace K] →
+      ∀ (f : K → X), Continuous f → IsOpen (f ⁻¹' s)) → IsOpen s) :
+    KSpace X where
+  restrictGenTopology := {
+    isOpen_of_forall_induced := by
+      intro s hs
+      apply h s
+      intro K _ _ f hf
+      change IsOpen ((fun x ↦ ⟨f x, mem_image_of_mem f trivial⟩) ⁻¹' ((f '' univ) ↓∩ s))
+      apply (hf.subtype_mk fun x ↦ mem_image_of_mem f trivial).isOpen_preimage
+      exact hs (f '' univ) (CompactSpace.isCompact_univ.image hf)}
+
+/-- Every compactly generated space is a k-space. -/
+instance kspace_of_compactlyGeneratedSpace {X : Type*} [TopologicalSpace X]
+    [CompactlyGeneratedSpace X] : KSpace X := by
+  apply kspace_of_preimage_isOpen
+  intro s h
+  apply CompactlyGeneratedSpace.isOpen'
+  intro K _ _ _ f hf
+  exact h K f hf
+
+/-- A k-space that is Hausdorff is compactly generated. -/
+instance compactlyGeneratedSpace_of_kspace_of_t2 {X : Type*} [TopologicalSpace X] [T2Space X]
+    [KSpace X] : CompactlyGeneratedSpace X := by
+  apply compactlyGeneratedSpace_of_isClosed_of_t2
+  intro s hs
+  rw [isClosed_iff]
+  intro K hK
+  rw [← Subtype.preimage_coe_inter_self]
+  exact (hs K hK).preimage_val
+
+/-- A type synonym used for the k-ification of a topological space.-/
+def kification (X : Type*) := X
+
+/--For a topological space `X` the k-ification is defined as:
+`A` is open iff for all compact sets `B`, the intersection `A ∩ B` is open in `B`.-/
+instance instkification {X : Type*} [TopologicalSpace X] : TopologicalSpace (kification X) where
+  IsOpen A := ∀ (K : Set X), IsCompact K → IsOpen (K ↓∩ A)
+  isOpen_univ := by simp
+  isOpen_inter A B hA hB K hK := (hA K hK).inter (hB K hK)
+  isOpen_sUnion s hs K hK := by
+    rw [preimage_val_sUnion]
+    apply isOpen_sUnion
+    rintro _ ⟨u, hu, rfl⟩
+    exact hs u hu K hK
+
+/-- A set `A` is the k-ification is open iff for all compact sets `K`,
+the intersection `K ∩ A` is open in `K`.-/
+lemma kification.isOpen_iff {X : Type*} [TopologicalSpace X] {A : Set (kification X)} :
+    IsOpen (X := kification X) A ↔
```

Comment:
Please don't abuse defeq `kification X = X`. E.g., `simp` in Lean 4 is very unhappy about this type of abuse. Instead, please define
```
def toKification : X ≃ kification X := Equiv.refl _
```
then use images or preimages under `toKification` and `toKification.symm` to move sets between the spaces. This also applied to the `TopologicalSpace` instance.